File: src/main/java/net/lomeli/lomlib/util/FluidUtil.java
package net.lomeli.lomlib.util;

import com.google.common.collect.Lists;

import java.util.Iterator;
import java.util.List;

import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;

import net.minecraftforge.fluids.*;
import net.minecraftforge.fml.common.registry.GameData;

public class FluidUtil {

    public static boolean isFluidWater(Fluid fluid) {
        return FluidRegistry.WATER.equals(fluid);
    }

    public static boolean isFluidWater(FluidStack fluid) {
        return fluid == null ? false : isFluidWater(fluid.getFluid());
    }

    public static boolean isFluidLava(Fluid fluid) {
        return FluidRegistry.LAVA.equals(fluid);
    }

    public static boolean isFluidLava(FluidStack fluid) {
        return fluid == null ? false : isFluidLava(fluid.getFluid());
    }

    public static boolean areFluidsEqual(FluidStack baseFluid, FluidStack secondFluid) {
        return (baseFluid != null && secondFluid != null) ? baseFluid.isFluidEqual(secondFluid) : false;
    }

    public static void combineFluidStacks(FluidStack baseFluid, FluidStack secondFluid) {
        if (baseFluid != null && secondFluid != null) {
            if (areFluidsEqual(baseFluid, secondFluid)) {
                baseFluid.amount += secondFluid.amount;
            }
        }
    }

    public static FluidStack getContainerStack(ItemStack stack) {
        if (stack != null) {
            return FluidContainerRegistry.isFilledContainer(stack) ? FluidContainerRegistry.getFluidForFilledItem(stack) :
                    (stack.getItem() instanceof IFluidContainerItem) ? ((IFluidContainerItem) stack.getItem()).getFluid(stack) : null;
        }
        return null;
    }

    public static Fluid getContainerFluid(ItemStack stack) {
        if (stack != null) {
            FluidStack fluidStack = getContainerStack(stack);
            return fluidStack != null ? fluidStack.getFluid() : null;
        }
        return null;
    }

    public static boolean isFilledContainer(ItemStack stack) {
        if (stack != null && stack.getItem() != null) {
            if (stack.getItem() instanceof IFluidContainerItem) {
                FluidStack fluidStack = ((IFluidContainerItem) stack.getItem()).getFluid(stack);
                if (fluidStack != null && fluidStack.getFluid() != null && fluidStack.amount > 0)
                    return true;
            }
            return FluidContainerRegistry.isFilledContainer(stack);
        }
        return false;
    }

    public static ItemStack getEmptyContainer(ItemStack stack) {
        if (stack != null && isFilledContainer(stack)) {
            if (stack.getItem() instanceof IFluidContainerItem) {
                ((IFluidContainerItem) stack.getItem()).drain(stack, Integer.MAX_VALUE, true);
                return stack;
            } else
                return FluidContainerRegistry.drainFluidContainer(stack);
        }
        return null;
    }

    public static boolean isFluidContainer(ItemStack stack) {
        if (stack == null || stack.getItem() == null) return false;
        if (FluidContainerRegistry.isContainer(stack))
            return true;
        if (stack.getItem() instanceof IFluidContainerItem)
            return true;
        return false;
    }

    public static int getStackCapacity(ItemStack stack) {
        if (stack.getItem() instanceof IFluidContainerItem)
            return ((IFluidContainerItem) stack.getItem()).getCapacity(stack);
        return FluidContainerRegistry.getContainerCapacity(stack);
    }

    public static void fillStack(ItemStack stack, FluidStack fluid) {
        if (stack == null || stack.getItem() == null) return;
        if (stack.getItem() instanceof IFluidContainerItem) {
            ((IFluidContainerItem) stack.getItem()).fill(stack, fluid, true);
        } else
            FluidContainerRegistry.fillFluidContainer(fluid, stack);
    }

    public static List<ItemStack> getContainersForFluid(Fluid targetFluid) {
        List<ItemStack> list = Lists.newArrayList();

        if (targetFluid != null) {
            for (FluidContainerRegistry.FluidContainerData data : FluidContainerRegistry.getRegisteredFluidContainerData()) {
                Fluid fluid = data.fluid.getFluid();
                if (fluid != null) {
                    if (fluid.equals(targetFluid) || fluid.getID() == targetFluid.getID())
                        list.add(data.filledContainer);
                }
            }
            Iterator<Item> itemIterator = GameData.getItemRegistry().iterator();
            if (itemIterator != null) {
                while (itemIterator.hasNext()) {
                    Item item = itemIterator.next();
                    if (item instanceof IFluidContainerItem) {
                        for (int i = 0; i <= item.getMaxDamage(); i++) {
                            ItemStack stack = new ItemStack(item, 1, i);
                            if (((IFluidContainerItem) item).fill(stack, new FluidStack(targetFluid, FluidContainerRegistry.BUCKET_VOLUME), true) == 0)
                                break;
                            FluidStack fluidStack = ((IFluidContainerItem) item).getFluid(stack);
                            if (fluidStack != null && fluidStack.getFluid() != null) {
                                Fluid fluid = fluidStack.getFluid();
                                if (fluid.equals(targetFluid) || fluid.getID() == targetFluid.getID())
                                    list.add(stack);
                            }
                        }
                    }
                }
            }
        }

        return list;
    }
}

File: src/main/kotlin/net/lomeli/lomlib/util/forge/FluidUtil.kt
package net.lomeli.lomlib.util.forge

import com.google.common.collect.Lists
import net.minecraft.init.Items
import net.minecraft.item.Item
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.util.EnumFacing
import net.minecraftforge.fluids.*
import net.minecraftforge.fluids.capability.CapabilityFluidHandler
import net.minecraftforge.fluids.capability.templates.FluidHandlerItemStack

//TODO Probably broken, needs fix
object FluidUtil {

    fun isFluidWater(fluid: Fluid): Boolean = FluidRegistry.WATER == fluid

    fun isFluidWater(fluid: FluidStack?): Boolean = if (fluid == null) false else isFluidWater(fluid.fluid)

    fun isFluidLava(fluid: Fluid): Boolean = FluidRegistry.LAVA == fluid

    fun isFluidLava(fluid: FluidStack?): Boolean = if (fluid == null) false else isFluidLava(fluid.fluid)

    fun areFluidsEqual(baseFluid: FluidStack?, secondFluid: FluidStack?): Boolean = if (baseFluid != null && secondFluid != null) baseFluid.isFluidEqual(secondFluid) else false

    fun combineFluidStacks(baseFluid: FluidStack?, secondFluid: FluidStack?) {
        if (baseFluid != null && secondFluid != null) {
            if (areFluidsEqual(baseFluid, secondFluid)) {
                baseFluid.amount += secondFluid.amount
            }
        }
    }

    fun getContainerStack(stack: ItemStack?): FluidStack? {
        if (!stack!!.isEmpty && stack.hasCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, EnumFacing.NORTH)) {
            var capability = stack.getCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, EnumFacing.NORTH) as FluidHandlerItemStack
            if (capability is FluidHandlerItemStack)
                return capability.fluid
        }
        return null
    }

    fun getContainerFluid(stack: ItemStack?): Fluid? {
        if (!stack!!.isEmpty) {
            val fluidStack = getContainerStack(stack)
            return fluidStack?.fluid
        }
        return null
    }

    fun isFilledContainer(stack: ItemStack?): Boolean {
        if (!stack!!.isEmpty && stack.hasCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, EnumFacing.NORTH)) {
            var capability = stack.getCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, EnumFacing.NORTH) as FluidHandlerItemStack
            return capability.drain(1, false) != null
        }
        return false
    }

    fun getEmptyContainer(stack: ItemStack?): ItemStack? {
        if (isFilledContainer(stack!!)) {
            if (stack.item === Items.LAVA_BUCKET || stack.item === Items.WATER_BUCKET || stack.item === Items.MILK_BUCKET)
                return ItemStack(Items.BUCKET)
            /*else if (stack.item is IFluidContainerItem) {
                (stack.item as IFluidContainerItem).drain(stack, (stack.item as IFluidContainerItem).getCapacity(stack), true)
                return stack
            } else
                return FluidContainerRegistry.drainFluidContainer(stack)*/
        }
        return null
    }

    fun isFluidContainer(stack: ItemStack?): Boolean {
        if (stack!!.isEmpty) return false
        return stack.hasCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, EnumFacing.NORTH)
    }

    fun getStackCapacity(stack: ItemStack): Int {
        if (stack.hasCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, EnumFacing.NORTH)) {
            var capability = stack.getCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, EnumFacing.NORTH)
            return capability!!.tankProperties[0].capacity
        }
        return -1
    }

    fun fillStack(stack: ItemStack?, fluid: FluidStack) {
        if (stack!!.isEmpty) return
        if (stack.hasCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, EnumFacing.NORTH)) {
            var capability = stack.getCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, EnumFacing.NORTH)
            capability!!.fill(fluid, true)
            return
        }
    }

    fun getContainersForFluid(targetFluid: Fluid?): List<ItemStack> {
        val list = Lists.newArrayList<ItemStack>()

        if (targetFluid != null) {
            val itemIterator = Item.REGISTRY.iterator()
            if (itemIterator != null) {
                while (itemIterator.hasNext()) {
                    val item = itemIterator.next()
                    val stackItem = ItemStack(item)
                    stackItem.tagCompound = NBTTagCompound()
                    if (stackItem.hasCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, EnumFacing.NORTH)) {
                        var capability = stackItem.getCapability(CapabilityFluidHandler.FLUID_HANDLER_CAPABILITY, EnumFacing.NORTH)
                        if (capability is FluidHandlerItemStack) {
                            val fluidStack = capability.fluid
                            if (fluidStack?.fluid == targetFluid || fluidStack?.fluid?.name == targetFluid.name) {
                                list.add(stackItem)
                                continue
                            }
                        }
                    }
                }
            }
        }

        return list
    }
}
